cURL function revision and date sorting addition

jessLundie opened the pull request:
Revised base cURL function and added functionality to alert for the number of past-due issues.

jessLundie committed:
cURL function revision and date sorting addition

Revised base cURL function and added functionality to alert for the number of past-due issues.

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
 	// Close cURL
 	curl_close( $process );
 
+	$result = json_decode( $result, true ); // Convert response to array
```
here's a small note about comments.

comments are _always great_ right?
well, comments, like code, have different value and there are different approaches to leaving them.

what I think happened here is that you were documenting your _process_ of understanding the problem, of how you came to solve it.

this is fun and tells a good story, but there's something in this comment that's _confusing_. this comment is _misleading_ because it asserts some behavior not obvious, not even inferable, from the code itself.

this lines absolutely _does not_ "convert [the] response to [an] array." it's possible that this could happen, if the response is the right shape, but other things can happen too given the same code and a different response.

```php
json_decode( "true", true ) === true;
json_decode( "3.141", true ) === 3.141;
json_decode( "null", true ) === NULL
```

These counter-examples contradict the comment and that's what I wanted to pause on to reflect. I love sharing the _why_ of code or the surprises of what's going on as a way to help someone new orient themselves here, but we can do better with this comment.

In a simple case like this nobody's going to go astray, but it's a common-enough type of situation to be a learning moment. Let's make a few observations:

 - the _code_ itself shows that what we are doing is decoding a JSON string. we might have to know a little about JSON and a little about `json_decode()` to know what's going on, but those are parts of the base knowledge we can assume people have to learn anyway.
 - "array" means different things in PHP than in JavaScript/JSON. in JSON it's always a list of things but in PHP it's _either_ a list _or_ a "map" (an associative array).
 - we're clearly expecting a JSON _map_ in the response.

A way we can change this comment is:
 - take it out since the code tells "what" we are doing
 - explain "why" we are doing it with something like ` // the response should be a JSON map`
 - explain fully in code with an additional constraint/check

If we want to make it clear what we're doing and we don't want to worry about the check we can add more context and safety with runtime assertions. This is probably more than we need to worry about here.

jessLundie replied:
Thank you! This makes so much sense.

``` 
If we want to make it clear what we're doing and we don't want to worry about the check we can add more context and safety with runtime assertions. 
```

Could you clarify what you mean here by runtime assertions? Or is that the bit that's more than we need to worry about here?

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
-	$url = 'https://api.github.com/users/jessLundie'; // Set source URL
+$media_header = array(
+	"Accept: application/vnd.github.inertia-preview+json",
+); // Custom media header required for projects in beta
```
this is a great comment because it explains why the value is set the way it is

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
+); // Custom media header required for projects in beta
 
-	$column_id = "15445284"; // Column id, found in the column link here: https://d.pr/i/W7RQJB
+$user_agent = "jessLundie"; // Set user agent
```
we know from the code itself that we are setting the `user_agent`, but I'm left asking "why is `jessLundie` the User Agent string?" and "why are we setting this?"

the comment here could be a great opportunity to make that case. it could be too that from the outside looking in, we don't care at all that this is communicated to GitHub via the User Agent. what matters _from the outside_? that's the name we want to use _as our interface_.

possibly, `$username = "jessLundie";`

jessLundie replied:
Ah, I see. From the outside, I think what matters is that this value is set to the account where the project you're trying to access lives. In this case, it could be a GitHub username or an organization name, for an organization-owned project. 

In this case, maybe just:

`$user = "jessLundie"; // Project owner - GH username or organization`

dmsnell replied:
or even `$project_owner = "jessLundie"; // GH username or organization`

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
+$user_agent = "jessLundie"; // Set user agent
 
-	$project_url = "https://api.github.com/projects/columns/" . $column_id . "/cards"; // Set project URL
+$token = ""; // Basic Auth Token
```
the comment could say where to find this

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
-	$headers = array(
-		"Accept: application/vnd.github.inertia-preview+json",
-	);
+function curl_fetch( $url, $media_header, $user_agent, $token, $api_url ) {
```
we happen to be using `curl` here but that's not _required_ or _inherent_ to what we're doing. can you think of a name that communicates only what we're doing and what matters to someone using this function? otherwise, what situation would we be in if we decided to use something other than `curl`? would we have to rewrite our existing code that calls this and make it call `some_other_fetch()`?

dmsnell commented on t51-projects/gh-project-board.php:
```diff
@@ -1,45 +1,70 @@
+
+	$label = curl_fetch( $url, $media_header, $user_agent, $token, $label_url  ); // API call to fetch labels
+
+	$label_list = array_merge( $label_list, $label ); // Add each label to list of labels
```
although I applaud the functional way you are appending the labels to the list, we're in a kind of mixed territory here. we're swapping out the value of `$label_list` so we can't just know what its value is without trying to run the code in our heads, but we're picking an immutable joiner function whose performance is well behind the more normative _mutable_ code.

when creating a list like this in PHP by appending elements we normally use the `[] = ` syntax.

```php
$label_list[] = $label;
```

were we to want to use a functional/immutable approach we could, though it would be a little more awkward in PHP than it would be in JS.

```php
$label_list = array_reduce( $cards, function( $labels, $card ) {
    $label_url = …
    $label = curl_fetch( … );

    return array_merge( $labels, $label );
}, [] );
```

this version never _reassigns_ the value of `$label_list`. nonetheless, your PHPeers will likely prefer the `$label_list[] = $label` form

jessLundie replied:
Changed! But I'm not sure I understand the mutable / immutable distinction in this example. Can we discuss this more tomorrow?

jessLundie committed:
Added date check for year turnover + comments

Addressed feedback on comments, and added date check to handle Quarter 4 - Quarter 1 turnover
